Fixes #22371, #22563 - Add pagination and filtering to RH repos

tstrachota opened the pull request:
- Adds a wrapper around `patternfly-react`'s paginator. It provides default pagination values and a simplified callback.
- Changes right pane on the RH repos page to call repositories API instead of querying repo sets and mapping it to repos. I had to do it to get correct numbers for pagination.
- Adds a new flag to repos scoped search for filtering only Red Hat repos.
- Repo content type filter.

Please note that this PR requires latest scoped-search to get search working (https://github.com/wvanbergen/scoped_search/commit/ec032a1434567de1824fea0b629a30694cc2194b)

jlsherrill commented on app/models/katello/repository.rb:
```diff
@@ -742,6 +743,19 @@ def self.linked_repositories
+      value = value == 'true'
+      value = !value if operator == '<>'
+
+      product_ids = Katello::Product.redhat.pluck(:id)
```
can you use select here so that we aren't issuing a second query?

tstrachota replied:
Not sure I understand @jlsherrill. Are you suggesting to use
```ruby
product_ids = Katello::Product.redhat.select(:id).map { |p| p.id }
```
instead? They generate the same query.

jlsherrill commented on app/models/katello/repository.rb:
```diff
@@ -165,6 +165,7 @@ class Repository < Katello::Model
     scoped_search :on => :distribution_bootable, :complete_value => true
     scoped_search :on => :distribution_uuid, :complete_value => true
     scoped_search :on => :ignore_global_proxy, :complete_value => true
+    scoped_search :on => :redhat, :complete_value => { :true => true, :false => false }, :ext_method => :search_by_redhat
```
mind adding a test for this search?

tstrachota replied:
Done, see `model/repository_test.rb`

waldenraines commented on webpack/components/PaginationRow/index.js:
```diff
@@ -0,0 +1,88 @@
+  const pagination = props.pagination || {};
+  const defaultPagination = {
+    page: 1,
+    perPage: 10,
```
This should probably be hooked up to the user setting. I am fine with handling this in another PR.

waldenraines commented on webpack/components/PaginationRow/index.js:
```diff
@@ -0,0 +1,88 @@
+  return { ...defaultPagination, ...pagination };
+};
+
+class PaginationRow extends Component {
```
Do you mind adding tests for this class?

tstrachota replied:
Done

waldenraines commented:
The per page selector looks a little odd to me.

![screen shot 2018-03-01 at 9 54 35 am](https://user-images.githubusercontent.com/4116405/36851624-a0e8bd3e-1d37-11e8-8cc7-750d81f5ec62.png)

waldenraines commented:
The per page selector looks a little odd to me because it's not vertically aligned with the rest of the pagination.

![screen shot 2018-03-01 at 9 54 35 am](https://user-images.githubusercontent.com/4116405/36851624-a0e8bd3e-1d37-11e8-8cc7-750d81f5ec62.png)

Removing the `.btn-group` class on the dropup seemed to fix it.

waldenraines commented:
[test katello]

waldenraines commented:
[test katello]

tstrachota commented:
Updated. I also added code that hooks up the content type filter dropdown. Both changes are quite connected.

@waldenraines it seems that wrong position of the text is caused by missing styles for `.pagination-pf-pagesize`. They're included in patternfly and everything works in patternfly-react storybook. I haven't found what's wrong here.

tstrachota commented:
Please note that it requires latest scoped-search to get search working (https://github.com/wvanbergen/scoped_search/commit/ec032a1434567de1824fea0b629a30694cc2194b)

waldenraines commented:
> @waldenraines it seems that wrong position of the text is caused by missing styles for .pagination-pf-pagesize. They're included in patternfly and everything works in patternfly-react storybook. I haven't found what's wrong here.

I have a bunch of styling updates to make from some feedback given to me by @rohoover.  I can take a look at this as part of those changes.

waldenraines commented:
@tstrachota it looks like the content type selector is applying to both enabled and available regardless of what is selected in the Available/Enabled/Both selector.  The interaction as defined by @Rohoover is that the content type dropdown should update the results based on what is selected in the Available/Enabled/Both selector.  Sorry if that wasn't clear.

tstrachota commented:
I see, sorry for that. I've updated the code and it should work as expected now.

I moved the state for search and type filtering one component up to `SearchBar`. I think it improved clarity of the code.

waldenraines commented:
@tstrachota looks like some tests or snapshots need updating for tests to pass

jturel commented:
@tstrachota sorry to revive this old PR but I have a question regarding the switch made here from /repository_sets to /repositories

You mentioned something about the paging. Can you recall specifically what it was that made you switch the API?

tstrachota commented:
@jturel AFAIR the original solution used to make a request on repo sets index (returns also repositories for each repo set), take out repositories from that response and glue them into one list. It was impossible to implement pagination for repos this way because it was actually paging repo sets. The result was different number of repos per page and that was extremely unintuitive UX.
